verticapy/learn/causal_model.py

```diff
@@ -0,0 +1,343 @@
+							deleted_edges.add((x,y))
+							skeleton_edges[x].remove(y)
+							skeleton_edges[y].remove(x)
+							sepsets[frozenset((x,y))] = z
+							break
+				all_pairs = all_pairs - deleted_edges
+		
+			elif method == 'orig':
+				deleted_edges = set()
+				for (x,y) in all_pairs:                
+					for z in itertools.chain(
+											 itertools.combinations(skeleton_edges[x]-set([y]), cond_set_size),
+											 itertools.combinations(skeleton_edges[y]-set([x]), cond_set_size)): # set of all conditionals
+						z = list(z)
+						if cit(self.vdf, x, y, z, alpha = significance_level):
+							deleted_edges.add((x,y))
+							skeleton_edges[x].remove(y)
+							skeleton_edges[y].remove(x)
+							sepsets[frozenset((x,y))] = z
+							break
+				all_pairs = all_pairs - deleted_edges
+			else:
+				raise AttributeError("Choose from valid method")
+			
+			if show_progress:
+				progress_bar.update(1)
+				progress_bar.set_description(f"Using n conditional variables: {cond_set_size + 1}")
+		
+		if show_progress:
+			progress_bar.close()
+
+		self.skeleton = skeleton_edges
+		self.sepsets = sepsets
+
+		return (skeleton_edges, sepsets)
+
+	# ---#
+	def build_pdag(self):
+		"""
+		Builds the partially directed acyclic graph
+
+		Returns
+		-------
+		pdag: dict
+			directed egdes in an adjacency dictionary
+		"""
+		skeleton_edges = self.skeleton
+		sepsets = self.sepsets   
+		
+		# Create partially directed graph
+		columns = skeleton_edges.keys()	
+		pdag = {x: skeleton_edges[x].copy() for x in columns}
+		
+		# 1 Orientate the triples X - Z - Y into X -> Z <- Y  
+		node_pairs = set(itertools.permutations(columns, 2))
+
+		for (x,y) in node_pairs:
+			if y not in skeleton_edges[x]:
+				for z in set(skeleton_edges[x] & skeleton_edges[y]):
+					#if z in pdag[x] and z in pdag[y] and x in pdag[z] and y in pdag[z]:
+					if z not in sepsets[frozenset((x,y))]:
+						if x in pdag[z]:
+							pdag[z].remove(x)
+						if y in pdag[z]:
+							pdag[z].remove(y)
+		
+		edges_to_be_oriented = True
+		while edges_to_be_oriented:
+			remaining_edges = sum([len(pdag[x]) for x in columns])
+		# 2 Orientate non-colliders X -> Z - Y into X -> X -> Y
+			for (x,z) in node_pairs:
+				if z in pdag[x] and x not in pdag[z]:
+					for y in pdag[z]:
+						if z in pdag[y] and y not in skeleton_edges[x]:
+							pdag[y].remove(z)             
+		# 3 Orientate non cycles X-Y with a directed path from X to Y orienate X->Y
+			for (x,y) in node_pairs:
+				if x in pdag[y] and y in pdag[x]:
+					if (is_dir_path(pdag, x, y)):
+						pdag[y].remove(x)   
+		# 4 For each X - Z - Y with X -> W, Y -> W and Z-W orienate Z->W
+			for (z,w) in node_pairs:
+				if w in pdag[z] and z in pdag[w]:
+					for (x,y) in itertools.combinations(pdag[z], 2):
+						if z in pdag[x] and z in pdag[y] and w in pdag[x] and w in pdag[y] and x not in pdag[w] and y not in pdag[w]:
+							pdag[w].remove(z)
+			edges_to_be_oriented = remaining_edges > sum([len(pdag[x]) for x in columns])
```

Comment:
I don't understand the logic of this line. Can you please briefly explain it in a short comment?

Reply:
If no edges got removed this iteration then the next iteration no edges will also get removed so the function can stop early and we can identify no edges got removed when `remaining_edges = sum([len(pdag[x]) for x in columns])` because `remaining_edges` is the count before this iteration and `sum([len(pdag[x]) for x in columns])` is the count after this iteration.